backend/glp/service.py

```
@@ -0,0 +1,684 @@
+
+
+def _lock_program():
+    queryset = GLPProgram.objects.select_for_update(
+        nowait=connection.features.has_select_for_update_nowait,
+    )
+    try:
+        return queryset.get(pk=1)
+    except GLPProgram.DoesNotExist:
+        raise GateError('GLP program does not exist yet.')
+    except DatabaseError:
+        raise RunInProgress('Another GLP operation holds the program lock.')
```

Ruff (0.16.1)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)